BUILD: -Wall causing fits for GCC 8.3.0-6

`error: ‘char* strncpy(char*, const char*, size_t)’ output may be truncated copying 15 bytes from a string of length 15 [-Werror=stringop-truncation]
   strncpy(szReport, szTemp, 15);`

Any way to get the default build to negotiate this one?

Just tested now. Doesn't appear to have an impact. Still fails with:

```
oiio/src/libutil/SHA1.cpp:286:10: error: ‘char* strncpy(char*, const char*, size_t)’ output may be truncated copying 15 bytes from a string of length 15 [-Werror=stringop-truncation]
   strncpy(szReport, szTemp, 15);

```

--- a/src/include/OpenImageIO/SHA1.h
+++ b/src/include/OpenImageIO/SHA1.h
@@ -1,81 +1,109 @@
 /*
-	100% free public domain implementation of the SHA-1 algorithm
-	by Dominik Reichl <dominik.reichl@t-online.de>
-	Web: http://www.dominik-reichl.de/
+  100% free public domain implementation of the SHA-1 algorithm
+  by Dominik Reichl <dominik.reichl@t-online.de>
+  Web: http://www.dominik-reichl.de/
 
-	Version 1.8 - 2008-03-16
-	- Converted project files to Visual Studio 2008 format.
-	- Added Unicode support for HashFile utility method.
-	- Added support for hashing files using the HashFile method that are
-	  larger than 2 GB.
-	- HashFile now returns an error code instead of copying an error
-	  message into the output buffer.
-	- GetHash now returns an error code and validates the input parameter.
-	- Added ReportHashStl STL utility method.
-	- Added REPORT_HEX_SHORT reporting mode.
-	- Improved Linux compatibility of test program.
+  Version 2.1 - 2012-06-19
+  - Deconstructor (resetting internal variables) is now only
+    implemented if SHA1_WIPE_VARIABLES is defined (which is the
+    default).
+  - Renamed inclusion guard to contain a GUID.
+  - Demo application is now using C++/STL objects and functions.
+  - Unicode build of the demo application now outputs the hashes of both
+    the ANSI and Unicode representations of strings.
+  - Various other demo application improvements.
 
-	Version 1.7 - 2006-12-21
-	- Fixed buffer underrun warning that appeared when compiling with
-	  Borland C Builder (thanks to Rex Bloom and Tim Gallagher for the
-	  patch).
-	- Breaking change: ReportHash writes the final hash to the start
-	  of the buffer, i.e. it's not appending it to the string anymore.
-	- Made some function parameters const.
-	- Added Visual Studio 2005 project files to demo project.
+  Version 2.0 - 2012-06-14
+  - Added 'limits.h' include.
+  - Renamed inclusion guard and macros for compliancy (names beginning
+    with an underscore are reserved).
 
-	Version 1.6 - 2005-02-07 (thanks to Howard Kapustein for patches)
-	- You can set the endianness in your files, no need to modify the
-	  header file of the CSHA1 class anymore.
-	- Aligned data support.
-	- Made support/compilation of the utility functions (ReportHash and
-	  HashFile) optional (useful when bytes count, for example in embedded
-	  environments).
+  Version 1.9 - 2011-11-10
+  - Added Unicode test vectors.
+  - Improved support for hashing files using the HashFile method that
+    are larger than 4 GB.
+  - Improved file hashing performance (by using a larger buffer).
+  - Disabled unnecessary compiler warnings.
+  - Internal variables are now private.
 
-	Version 1.5 - 2005-01-01
-	- 64-bit compiler compatibility added.
-	- Made variable wiping optional (define SHA1_WIPE_VARIABLES).
-	- Removed unnecessary variable initializations.
-	- ROL32 improvement for the Microsoft compiler (using _rotl).
+  Version 1.8 - 2009-03-16
+  - Converted project files to Visual Studio 2008 format.
+  - Added Unicode support for HashFile utility method.
+  - Added support for hashing files using the HashFile method that are
+    larger than 2 GB.
+  - HashFile now returns an error code instead of copying an error
+    message into the output buffer.
+  - GetHash now returns an error code and validates the input parameter.
+  - Added ReportHashStl STL utility method.
+  - Added REPORT_HEX_SHORT reporting mode.
+  - Improved Linux compatibility of test program.
 
-	Version 1.4 - 2004-07-22
-	- CSHA1 now compiles fine with GCC 3.3 under MacOS X  (thanks to Larry
-	  Hastings).
+  Version 1.7 - 2006-12-21
+  - Fixed buffer underrun warning that appeared when compiling with
+    Borland C Builder (thanks to Rex Bloom and Tim Gallagher for the
+    patch).
+  - Breaking change: ReportHash writes the final hash to the start
+    of the buffer, i.e. it's not appending it to the string anymore.
+  - Made some function parameters const.
+  - Added Visual Studio 2005 project files to demo project.
 
-	Version 1.3 - 2003-08-17
-	- Fixed a small memory bug and made a buffer array a class member to
-	  ensure correct working when using multiple CSHA1 class instances at
-	  one time.
+  Version 1.6 - 2005-02-07 (thanks to Howard Kapustein for patches)
+  - You can set the endianness in your files, no need to modify the
+    header file of the CSHA1 class anymore.
+  - Aligned data support.
+  - Made support/compilation of the utility functions (ReportHash and
+    HashFile) optional (useful when bytes count, for example in embedded
+    environments).
 
-	Version 1.2 - 2002-11-16
-	- Borlands C++ compiler seems to have problems with string addition
-	  using sprintf. Fixed the bug which caused the digest report function
-	  not to work properly. CSHA1 is now Borland compatible.
+  Version 1.5 - 2005-01-01
+  - 64-bit compiler compatibility added.
+  - Made variable wiping optional (define SHA1_WIPE_VARIABLES).
+  - Removed unnecessary variable initializations.
+  - ROL32 improvement for the Microsoft compiler (using _rotl).
 
-	Version 1.1 - 2002-10-11
-	- Removed two unnecessary header file includes and changed BOOL to
-	  bool. Fixed some minor bugs in the web page contents.
+  Version 1.4 - 2004-07-22
+  - CSHA1 now compiles fine with GCC 3.3 under Mac OS X (thanks to Larry
+    Hastings).
 
-	Version 1.0 - 2002-06-20
-	- First official release.
+  Version 1.3 - 2003-08-17
+  - Fixed a small memory bug and made a buffer array a class member to
+    ensure correct working when using multiple CSHA1 class instances at
+    one time.
 
-	======== Test Vectors (from FIPS PUB 180-1) ========
+  Version 1.2 - 2002-11-16
+  - Borlands C++ compiler seems to have problems with string addition
+    using sprintf. Fixed the bug which caused the digest report function
+    not to work properly. CSHA1 is now Borland compatible.
 
-	SHA1("abc") =
-		A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
+  Version 1.1 - 2002-10-11
+  - Removed two unnecessary header file includes and changed BOOL to
+    bool. Fixed some minor bugs in the web page contents.
 
-	SHA1("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq") =
-		84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
+  Version 1.0 - 2002-06-20
+  - First official release.
 
-	SHA1(A million repetitions of "a") =
-		34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
+  ================ Test Vectors ================
+
+  SHA1("abc" in ANSI) =
+    A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
+  SHA1("abc" in Unicode LE) =
+    9F04F41A 84851416 2050E3D6 8C1A7ABB 441DC2B5
+
+  SHA1("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
+    in ANSI) =
+    84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
+  SHA1("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
+    in Unicode LE) =
+    51D7D876 9AC72C40 9C5B0E3F 69C60ADC 9A039014
+
+  SHA1(A million repetitions of "a" in ANSI) =
+    34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
+  SHA1(A million repetitions of "a" in Unicode LE) =
+    C4609560 A108A0C6 26AA7F2B 38A65566 739353C5
 */
-// clang-format off
 
-#ifndef ___SHA1_HDR___
-#define ___SHA1_HDR___
-
-#include <climits>
+#ifndef SHA1_H_A545E61D43E9404E8D736869AB3CBFE7
+#define SHA1_H_A545E61D43E9404E8D736869AB3CBFE7
 
 #include <OpenImageIO/export.h>
 #include <OpenImageIO/oiioversion.h>
@@ -93,10 +121,11 @@
 #endif
 
 #include <memory.h>
+#include <limits.h>
 
 #ifdef SHA1_UTILITY_FUNCTIONS
-#include <cstdio>
-#include <cstring>
+#include <stdio.h>
+#include <string.h>
 #endif
 
 #ifdef SHA1_STL_FUNCTIONS
@@ -122,28 +151,67 @@
 #define SHA1_WIPE_VARIABLES
 #endif
 
-#ifdef _MSC_VER
-#define _sntprintf std::snprintf
+#if defined(SHA1_HAS_TCHAR)
+#include <tchar.h>
 #else
-#define _sntprintf std::snprintf
+#ifdef _MSC_VER
+#include <tchar.h>
+#else
+#ifndef TCHAR
+#define TCHAR char
 #endif
-
-// Fallback, if no 64-bit support
-#ifndef _fseeki64
-#define _fseeki64 fseek
+#ifndef _T
+#define _T(__x) (__x)
+#define _tmain main
+#define _tprintf printf
+#define _getts gets
+#define _tcslen strlen
+#define _tfopen fopen
+#define _tcscpy strcpy
+#define _tcscat strcat
+#define _sntprintf snprintf
 #endif
-#ifndef _ftelli64
-#define _ftelli64 ftell
+#endif
 #endif
 
 ///////////////////////////////////////////////////////////////////////////
 // Define variable types
 
-#define UINT_8  uint8_t
-#define UINT_32 uint32_t
-#define UINT_64 uint64_t
-#define INT_64  int64_t
+#ifndef UINT_8
+#ifdef _MSC_VER // Compiling with Microsoft compiler
+#define UINT_8 unsigned __int8
+#else // !_MSC_VER
+#define UINT_8 unsigned char
+#endif // _MSC_VER
+#endif
 
+#ifndef UINT_32
+#ifdef _MSC_VER // Compiling with Microsoft compiler
+#define UINT_32 unsigned __int32
+#else // !_MSC_VER
+#if (ULONG_MAX == 0xFFFFFFFFUL)
+#define UINT_32 unsigned long
+#else
+#define UINT_32 unsigned int
+#endif
+#endif // _MSC_VER
+#endif // UINT_32
+
+#ifndef INT_64
+#ifdef _MSC_VER // Compiling with Microsoft compiler
+#define INT_64 __int64
+#else // !_MSC_VER
+#define INT_64 long long
+#endif // _MSC_VER
+#endif // INT_64
+
+#ifndef UINT_64
+#ifdef _MSC_VER // Compiling with Microsoft compiler
+#define UINT_64 unsigned __int64
+#else // !_MSC_VER
+#define UINT_64 unsigned long long
+#endif // _MSC_VER
+#endif // UINT_64
 
 ///////////////////////////////////////////////////////////////////////////
 // Declare SHA-1 workspace
@@ -160,7 +228,7 @@ class CSHA1
 {
 public:
 #ifdef SHA1_UTILITY_FUNCTIONS
-	// Different formats for ReportHash
+	// Different formats for ReportHash(Stl)
 	enum REPORT_TYPE
 	{
 		REPORT_HEX = 0,
@@ -171,8 +239,41 @@ public:
 
 	// Constructor and destructor
 	CSHA1();
-	~CSHA1();
 
+#ifdef SHA1_WIPE_VARIABLES
+	~CSHA1();
+#endif
+
+	void Reset();
+
+	// Hash in binary data and strings
+	void Update(const UINT_8* pbData, UINT_32 uLen);
+
+#ifdef SHA1_UTILITY_FUNCTIONS
+	// Hash in file contents
+	bool HashFile(const TCHAR* tszFileName);
+#endif
+
+	// Finalize hash; call it before using ReportHash(Stl)
+	void Final();
+
+#ifdef SHA1_UTILITY_FUNCTIONS
+	bool ReportHash(TCHAR* tszReport, REPORT_TYPE rtReportType = REPORT_HEX) const;
+#endif
+
+#ifdef SHA1_STL_FUNCTIONS
+	bool ReportHashStl(std::basic_string<TCHAR>& strOut, REPORT_TYPE rtReportType =
+		REPORT_HEX) const;
+#endif
+
+	// Get the raw message digest (20 bytes)
+	bool GetHash(UINT_8* pbDest20) const;
+
+private:
+	// Private SHA-1 transformation
+	void Transform(UINT_32* pState, const UINT_8* pBuffer);
+
+	// Member variables
 	UINT_32 m_state[5];
 	UINT_32 m_count[2];
 	UINT_32 m_reserved0[1]; // Memory alignment padding
@@ -180,39 +281,23 @@ public:
 	UINT_8 m_digest[20];
 	UINT_32 m_reserved1[3]; // Memory alignment padding
 
-	void Reset();
-
-	// Update the hash value
-	void Update(const UINT_8* pbData, UINT_32 uLen);
-
-#ifdef SHA1_UTILITY_FUNCTIONS
-	// Hash in file contents
-	bool HashFile(const char* szFileName);
-#endif
-
-	// Finalize hash, call before using ReportHash(Stl)
-	void Final();
-
-#ifdef SHA1_UTILITY_FUNCTIONS
-	bool ReportHash(char* szReport, REPORT_TYPE rtReportType = REPORT_HEX) const;
-#endif
-
-#ifdef SHA1_STL_FUNCTIONS
-	bool ReportHashStl(std::string& strOut, REPORT_TYPE rtReportType =
-		REPORT_HEX) const;
-#endif
-
-	bool GetHash(UINT_8* pbDest) const;
-
-private:
-	// Private SHA-1 transformation
-	void Transform(UINT_32* pState, const UINT_8* pBuffer);
-
-	// Member variables
 	UINT_8 m_workspace[64];
 	SHA1_WORKSPACE_BLOCK* m_block; // SHA1 pointer to the byte array above
 };
 
 OIIO_NAMESPACE_END
 
-#endif // ___SHA1_HDR___
+#ifndef DO_NOT_UNDEFINE_SHA1
+#undef TCHAR
+#undef _T
+#undef _tmain
+#undef _tprintf
+#undef _getts
+#undef _tcslen
+#undef _tfopen
+#undef _tcscpy
+#undef _tcscat
+#undef _sntprintf
+#endif
+
+#endif // SHA1_H_A545E61D43E9404E8D736869AB3CBFE7

--- a/src/libutil/SHA1.cpp
+++ b/src/libutil/SHA1.cpp
@@ -1,29 +1,29 @@
 /*
-	100% free public domain implementation of the SHA-1 algorithm
-	by Dominik Reichl <dominik.reichl@t-online.de>
-	Web: http://www.dominik-reichl.de/
+  100% free public domain implementation of the SHA-1 algorithm
+  by Dominik Reichl <dominik.reichl@t-online.de>
+  Web: http://www.dominik-reichl.de/
 
-	See header file for version history.
+  See header file for version history and test vectors.
 */
 
 // If compiling with MFC, you might want to add #include "StdAfx.h"
 // clang-format off
 
+#define DO_NOT_UNDEFINE_SHA1
+#define _CRT_SECURE_NO_WARNINGS
 #include <OpenImageIO/SHA1.h>
 #include <OpenImageIO/hash.h>
 #include <OpenImageIO/dassert.h>
 #include <OpenImageIO/strutil.h>
 
-#ifdef SHA1_UTILITY_FUNCTIONS
-#define SHA1_MAX_FILE_BUFFER 8000
-#endif
+#define SHA1_MAX_FILE_BUFFER (32 * 20 * 820)
 
-// Rotate x bits to the left
+// Rotate p_val32 by p_nBits bits to the left
 #ifndef ROL32
 #ifdef _MSC_VER
-#define ROL32(_val32,_nBits) _rotl(_val32,_nBits)
+#define ROL32(p_val32,p_nBits) _rotl(p_val32,p_nBits)
 #else
-#define ROL32(_val32,_nBits) (((_val32)<<(_nBits))|((_val32)>>(32-(_nBits))))
+#define ROL32(p_val32,p_nBits) (((p_val32)<<(p_nBits))|((p_val32)>>(32-(p_nBits))))
 #endif
 #endif
 
@@ -34,16 +34,21 @@
 #define SHABLK0(i) (m_block->l[i])
 #endif
 
-#define SHABLK(i) (m_block->l[i&15] = ROL32(m_block->l[(i+13)&15] ^ m_block->l[(i+8)&15] \
-	^ m_block->l[(i+2)&15] ^ m_block->l[i&15],1))
+#define SHABLK(i) (m_block->l[i&15] = ROL32(m_block->l[(i+13)&15] ^ \
+	m_block->l[(i+8)&15] ^ m_block->l[(i+2)&15] ^ m_block->l[i&15],1))
 
 // SHA-1 rounds
-#define _R0(v,w,x,y,z,i) {z+=((w&(x^y))^y)+SHABLK0(i)+0x5A827999+ROL32(v,5);w=ROL32(w,30);}
-#define _R1(v,w,x,y,z,i) {z+=((w&(x^y))^y)+SHABLK(i)+0x5A827999+ROL32(v,5);w=ROL32(w,30);}
-#define _R2(v,w,x,y,z,i) {z+=(w^x^y)+SHABLK(i)+0x6ED9EBA1+ROL32(v,5);w=ROL32(w,30);}
-#define _R3(v,w,x,y,z,i) {z+=(((w|x)&y)|(w&x))+SHABLK(i)+0x8F1BBCDC+ROL32(v,5);w=ROL32(w,30);}
-#define _R4(v,w,x,y,z,i) {z+=(w^x^y)+SHABLK(i)+0xCA62C1D6+ROL32(v,5);w=ROL32(w,30);}
+#define S_R0(v,w,x,y,z,i) {z+=((w&(x^y))^y)+SHABLK0(i)+0x5A827999+ROL32(v,5);w=ROL32(w,30);}
+#define S_R1(v,w,x,y,z,i) {z+=((w&(x^y))^y)+SHABLK(i)+0x5A827999+ROL32(v,5);w=ROL32(w,30);}
+#define S_R2(v,w,x,y,z,i) {z+=(w^x^y)+SHABLK(i)+0x6ED9EBA1+ROL32(v,5);w=ROL32(w,30);}
+#define S_R3(v,w,x,y,z,i) {z+=(((w|x)&y)|(w&x))+SHABLK(i)+0x8F1BBCDC+ROL32(v,5);w=ROL32(w,30);}
+#define S_R4(v,w,x,y,z,i) {z+=(w^x^y)+SHABLK(i)+0xCA62C1D6+ROL32(v,5);w=ROL32(w,30);}
 
+#ifdef _MSC_VER
+#pragma warning(push)
+// Disable compiler warning 'Conditional expression is constant'
+#pragma warning(disable: 4127)
+#endif
 
 OIIO_NAMESPACE_BEGIN
 
@@ -99,10 +104,12 @@ CSHA1::CSHA1()
 	Reset();
 }
 
+#ifdef SHA1_WIPE_VARIABLES
 CSHA1::~CSHA1()
 {
 	Reset();
 }
+#endif
 
 void CSHA1::Reset()
 {
@@ -123,27 +130,27 @@ void CSHA1::Transform(UINT_32* pState, const UINT_8* pBuffer)
 
 	memcpy(m_block, pBuffer, 64);
 
-	// 4 rounds of 20 operations each. Loop unrolled.
-	_R0(a,b,c,d,e, 0); _R0(e,a,b,c,d, 1); _R0(d,e,a,b,c, 2); _R0(c,d,e,a,b, 3);
-	_R0(b,c,d,e,a, 4); _R0(a,b,c,d,e, 5); _R0(e,a,b,c,d, 6); _R0(d,e,a,b,c, 7);
-	_R0(c,d,e,a,b, 8); _R0(b,c,d,e,a, 9); _R0(a,b,c,d,e,10); _R0(e,a,b,c,d,11);
-	_R0(d,e,a,b,c,12); _R0(c,d,e,a,b,13); _R0(b,c,d,e,a,14); _R0(a,b,c,d,e,15);
-	_R1(e,a,b,c,d,16); _R1(d,e,a,b,c,17); _R1(c,d,e,a,b,18); _R1(b,c,d,e,a,19);
-	_R2(a,b,c,d,e,20); _R2(e,a,b,c,d,21); _R2(d,e,a,b,c,22); _R2(c,d,e,a,b,23);
-	_R2(b,c,d,e,a,24); _R2(a,b,c,d,e,25); _R2(e,a,b,c,d,26); _R2(d,e,a,b,c,27);
-	_R2(c,d,e,a,b,28); _R2(b,c,d,e,a,29); _R2(a,b,c,d,e,30); _R2(e,a,b,c,d,31);
-	_R2(d,e,a,b,c,32); _R2(c,d,e,a,b,33); _R2(b,c,d,e,a,34); _R2(a,b,c,d,e,35);
-	_R2(e,a,b,c,d,36); _R2(d,e,a,b,c,37); _R2(c,d,e,a,b,38); _R2(b,c,d,e,a,39);
-	_R3(a,b,c,d,e,40); _R3(e,a,b,c,d,41); _R3(d,e,a,b,c,42); _R3(c,d,e,a,b,43);
-	_R3(b,c,d,e,a,44); _R3(a,b,c,d,e,45); _R3(e,a,b,c,d,46); _R3(d,e,a,b,c,47);
-	_R3(c,d,e,a,b,48); _R3(b,c,d,e,a,49); _R3(a,b,c,d,e,50); _R3(e,a,b,c,d,51);
-	_R3(d,e,a,b,c,52); _R3(c,d,e,a,b,53); _R3(b,c,d,e,a,54); _R3(a,b,c,d,e,55);
-	_R3(e,a,b,c,d,56); _R3(d,e,a,b,c,57); _R3(c,d,e,a,b,58); _R3(b,c,d,e,a,59);
-	_R4(a,b,c,d,e,60); _R4(e,a,b,c,d,61); _R4(d,e,a,b,c,62); _R4(c,d,e,a,b,63);
-	_R4(b,c,d,e,a,64); _R4(a,b,c,d,e,65); _R4(e,a,b,c,d,66); _R4(d,e,a,b,c,67);
-	_R4(c,d,e,a,b,68); _R4(b,c,d,e,a,69); _R4(a,b,c,d,e,70); _R4(e,a,b,c,d,71);
-	_R4(d,e,a,b,c,72); _R4(c,d,e,a,b,73); _R4(b,c,d,e,a,74); _R4(a,b,c,d,e,75);
-	_R4(e,a,b,c,d,76); _R4(d,e,a,b,c,77); _R4(c,d,e,a,b,78); _R4(b,c,d,e,a,79);
+	// 4 rounds of 20 operations each, loop unrolled
+	S_R0(a,b,c,d,e, 0); S_R0(e,a,b,c,d, 1); S_R0(d,e,a,b,c, 2); S_R0(c,d,e,a,b, 3);
+	S_R0(b,c,d,e,a, 4); S_R0(a,b,c,d,e, 5); S_R0(e,a,b,c,d, 6); S_R0(d,e,a,b,c, 7);
+	S_R0(c,d,e,a,b, 8); S_R0(b,c,d,e,a, 9); S_R0(a,b,c,d,e,10); S_R0(e,a,b,c,d,11);
+	S_R0(d,e,a,b,c,12); S_R0(c,d,e,a,b,13); S_R0(b,c,d,e,a,14); S_R0(a,b,c,d,e,15);
+	S_R1(e,a,b,c,d,16); S_R1(d,e,a,b,c,17); S_R1(c,d,e,a,b,18); S_R1(b,c,d,e,a,19);
+	S_R2(a,b,c,d,e,20); S_R2(e,a,b,c,d,21); S_R2(d,e,a,b,c,22); S_R2(c,d,e,a,b,23);
+	S_R2(b,c,d,e,a,24); S_R2(a,b,c,d,e,25); S_R2(e,a,b,c,d,26); S_R2(d,e,a,b,c,27);
+	S_R2(c,d,e,a,b,28); S_R2(b,c,d,e,a,29); S_R2(a,b,c,d,e,30); S_R2(e,a,b,c,d,31);
+	S_R2(d,e,a,b,c,32); S_R2(c,d,e,a,b,33); S_R2(b,c,d,e,a,34); S_R2(a,b,c,d,e,35);
+	S_R2(e,a,b,c,d,36); S_R2(d,e,a,b,c,37); S_R2(c,d,e,a,b,38); S_R2(b,c,d,e,a,39);
+	S_R3(a,b,c,d,e,40); S_R3(e,a,b,c,d,41); S_R3(d,e,a,b,c,42); S_R3(c,d,e,a,b,43);
+	S_R3(b,c,d,e,a,44); S_R3(a,b,c,d,e,45); S_R3(e,a,b,c,d,46); S_R3(d,e,a,b,c,47);
+	S_R3(c,d,e,a,b,48); S_R3(b,c,d,e,a,49); S_R3(a,b,c,d,e,50); S_R3(e,a,b,c,d,51);
+	S_R3(d,e,a,b,c,52); S_R3(c,d,e,a,b,53); S_R3(b,c,d,e,a,54); S_R3(a,b,c,d,e,55);
+	S_R3(e,a,b,c,d,56); S_R3(d,e,a,b,c,57); S_R3(c,d,e,a,b,58); S_R3(b,c,d,e,a,59);
+	S_R4(a,b,c,d,e,60); S_R4(e,a,b,c,d,61); S_R4(d,e,a,b,c,62); S_R4(c,d,e,a,b,63);
+	S_R4(b,c,d,e,a,64); S_R4(a,b,c,d,e,65); S_R4(e,a,b,c,d,66); S_R4(d,e,a,b,c,67);
+	S_R4(c,d,e,a,b,68); S_R4(b,c,d,e,a,69); S_R4(a,b,c,d,e,70); S_R4(e,a,b,c,d,71);
+	S_R4(d,e,a,b,c,72); S_R4(c,d,e,a,b,73); S_R4(b,c,d,e,a,74); S_R4(a,b,c,d,e,75);
+	S_R4(e,a,b,c,d,76); S_R4(d,e,a,b,c,77); S_R4(c,d,e,a,b,78); S_R4(b,c,d,e,a,79);
 
 	// Add the working vars back into state
 	pState[0] += a;
@@ -154,11 +161,10 @@ void CSHA1::Transform(UINT_32* pState, const UINT_8* pBuffer)
 
 	// Wipe variables
 #ifdef SHA1_WIPE_VARIABLES
-	a = b = c = d = e = 0;   // NOLINT
+	a = b = c = d = e = 0;
 #endif
 }
 
-// Use this function to hash in binary data and strings
 void CSHA1::Update(const UINT_8* pbData, UINT_32 uLen)
 {
 	UINT_32 j = ((m_count[0] >> 3) & 0x3F);
@@ -187,46 +193,34 @@ void CSHA1::Update(const UINT_8* pbData, UINT_32 uLen)
 }
 
 #ifdef SHA1_UTILITY_FUNCTIONS
-// Hash in file contents
-bool CSHA1::HashFile(const char* szFileName)
+bool CSHA1::HashFile(const TCHAR* tszFileName)
 {
-	if(szFileName == NULL) return false;
+	if(tszFileName == NULL) return false;
 
-#ifdef _WIN32
-    std::wstring wf = Strutil::utf8_to_utf16(szFileName);
-    FILE* fpIn = _wfopen(wf.c_str(), L"rb");
-#else
-	FILE* fpIn = fopen(szFileName, "rb");
-#endif
+	FILE* fpIn = _tfopen(tszFileName, _T("rb"));
 	if(fpIn == NULL) return false;
 
-	_fseeki64(fpIn, 0, SEEK_END);
-	const INT_64 lFileSize = _ftelli64(fpIn);
-	_fseeki64(fpIn, 0, SEEK_SET);
+	UINT_8* pbData = new UINT_8[SHA1_MAX_FILE_BUFFER];
+	if(pbData == NULL) { fclose(fpIn); return false; }
 
-	const INT_64 lMaxBuf = SHA1_MAX_FILE_BUFFER;
-	UINT_8 vData[SHA1_MAX_FILE_BUFFER];
-	INT_64 lRemaining = lFileSize;
-
-	while(lRemaining > 0)
+	bool bSuccess = true;
+	while(true)
 	{
-		const size_t uMaxRead = static_cast<size_t>((lRemaining > lMaxBuf) ?
-			lMaxBuf : lRemaining);
+		const size_t uRead = fread(pbData, 1, SHA1_MAX_FILE_BUFFER, fpIn);
 
-		const size_t uRead = fread(vData, 1, uMaxRead, fpIn);
-		if(uRead == 0)
+		if(uRead > 0)
+			Update(pbData, static_cast<UINT_32>(uRead));
+
+		if(uRead < SHA1_MAX_FILE_BUFFER)
 		{
-			fclose(fpIn);
-			return false;
+			if(feof(fpIn) == 0) bSuccess = false;
+			break;
 		}
-
-		Update(vData, static_cast<UINT_32>(uRead));
-
-		lRemaining -= static_cast<INT_64>(uRead);
 	}
 
 	fclose(fpIn);
-	return (lRemaining == 0);
+	delete[] pbData;
+	return bSuccess;
 }
 #endif
 
@@ -234,60 +228,60 @@ void CSHA1::Final()
 {
 	UINT_32 i;
 
-	UINT_8 finalcount[8];
+	UINT_8 pbFinalCount[8];
 	for(i = 0; i < 8; ++i)
-		finalcount[i] = (UINT_8)((m_count[((i >= 4) ? 0 : 1)]
-			>> ((3 - (i & 3)) * 8) ) & 255); // Endian independent
+		pbFinalCount[i] = static_cast<UINT_8>((m_count[((i >= 4) ? 0 : 1)] >>
+			((3 - (i & 3)) * 8) ) & 0xFF); // Endian independent
 
 	Update((UINT_8*)"\200", 1);
 
-	while ((m_count[0] & 504) != 448)
+	while((m_count[0] & 504) != 448)
 		Update((UINT_8*)"\0", 1);
 
-	Update(finalcount, 8); // Cause a SHA1Transform()
+	Update(pbFinalCount, 8); // Cause a Transform()
 
 	for(i = 0; i < 20; ++i)
-		m_digest[i] = (UINT_8)((m_state[i >> 2] >> ((3 - (i & 3)) * 8)) & 0xFF);
+		m_digest[i] = static_cast<UINT_8>((m_state[i >> 2] >> ((3 -
+			(i & 3)) * 8)) & 0xFF);
 
 	// Wipe variables for security reasons
 #ifdef SHA1_WIPE_VARIABLES
 	memset(m_buffer, 0, 64);
 	memset(m_state, 0, 20);
 	memset(m_count, 0, 8);
-	memset(finalcount, 0, 8);
+	memset(pbFinalCount, 0, 8);
 	Transform(m_state, m_buffer);
 #endif
 }
 
 #ifdef SHA1_UTILITY_FUNCTIONS
-// Get the final hash as a pre-formatted string
-bool CSHA1::ReportHash(char* szReport, REPORT_TYPE rtReportType) const
+bool CSHA1::ReportHash(TCHAR* tszReport, REPORT_TYPE rtReportType) const
 {
-	if(szReport == NULL) return false;
+	if(tszReport == NULL) return false;
 
-	char szTemp[16];
+	TCHAR tszTemp[16];
 
 	if((rtReportType == REPORT_HEX) || (rtReportType == REPORT_HEX_SHORT))
 	{
-		_sntprintf(szTemp, 15, "%02X", m_digest[0]);
-		strncpy(szReport, szTemp, 15);
+		_sntprintf(tszTemp, 15, _T("%02X"), m_digest[0]);
+		_tcscpy(tszReport, tszTemp);
 
-		const char* lpFmt = ((rtReportType == REPORT_HEX) ? " %02X" : "%02X");
+		const TCHAR* lpFmt = ((rtReportType == REPORT_HEX) ? _T(" %02X") : _T("%02X"));
 		for(size_t i = 1; i < 20; ++i)
 		{
-			_sntprintf(szTemp, 15, lpFmt, m_digest[i]);
-			strncat(szReport, szTemp, 15);
+			_sntprintf(tszTemp, 15, lpFmt, m_digest[i]);
+			_tcscat(tszReport, tszTemp);
 		}
 	}
 	else if(rtReportType == REPORT_DIGIT)
 	{
-		_sntprintf(szTemp, 15, "%u", m_digest[0]);
-		strncpy(szReport, szTemp, 15);
+		_sntprintf(tszTemp, 15, _T("%u"), m_digest[0]);
+		_tcscpy(tszReport, tszTemp);
 
 		for(size_t i = 1; i < 20; ++i)
 		{
-			_sntprintf(szTemp, 15, " %u", m_digest[i]);
-			strncat(szReport, szTemp, 15);
+			_sntprintf(tszTemp, 15, _T(" %u"), m_digest[i]);
+			_tcscat(tszReport, tszTemp);
 		}
 	}
 	else return false;
@@ -297,21 +291,24 @@ bool CSHA1::ReportHash(char* szReport, REPORT_TYPE rtReportType) const
 #endif
 
 #ifdef SHA1_STL_FUNCTIONS
-bool CSHA1::ReportHashStl(std::string& strOut, REPORT_TYPE rtReportType) const
+bool CSHA1::ReportHashStl(std::basic_string<TCHAR>& strOut, REPORT_TYPE rtReportType) const
 {
-	char szOut[84];
-	const bool bResult = ReportHash(szOut, rtReportType);
-	if(bResult) strOut = szOut;
+	TCHAR tszOut[84];
+	const bool bResult = ReportHash(tszOut, rtReportType);
+	if(bResult) strOut = tszOut;
 	return bResult;
 }
 #endif
 
-// Get the raw message digest
-bool CSHA1::GetHash(UINT_8* pbDest) const
+bool CSHA1::GetHash(UINT_8* pbDest20) const
 {
-	if(pbDest == NULL) return false;
-	memcpy(pbDest, m_digest, 20);
+	if(pbDest20 == NULL) return false;
+	memcpy(pbDest20, m_digest, 20);
 	return true;
 }
+
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
 
 OIIO_NAMESPACE_END

